File: packages/storage/src/lib/key-utils.ts
import { createHash, generateKeyPairSync, createPrivateKey, createPublicKey, KeyObject } from 'node:crypto';
import * as fs from 'fs/promises';
import { pathExists } from 'node-utils';
import { IStorageOptions } from './storage-factory';
import { ensureParentDirectoryExists } from 'node-utils';
import { FatalError } from 'utils';
import type { IPrivateKeyMap } from './encryption-types';

/**
 * Interface for key pair
 */
export interface IKeyPair {
    publicKey: KeyObject;
    privateKey: KeyObject;
}

/**
 * Generate a new RSA key pair
 * 
 * @returns A promise that resolves to the generated key pair
 */
export function generateKeyPair(): IKeyPair {
    const { publicKey, privateKey } = generateKeyPairSync('rsa', {
        modulusLength: 4096,
        publicKeyEncoding: {
            type: 'spki',
            format: 'pem'
        },
        privateKeyEncoding: {
            type: 'pkcs8',
            format: 'pem'
        }
    });

    return {
        publicKey: createPublicKey(publicKey),
        privateKey: createPrivateKey(privateKey)
    };
}

/**
 * Save a key pair to files
 * 
 * @param keyPair The key pair to save
 * @param keyFilePath Path to save the private key
 * @returns A promise that resolves when the keys are saved
 */
export async function saveKeyPair(keyPair: IKeyPair, keyFilePath: string): Promise<void> {
    const privateKeyPem = keyPair.privateKey.export({
        type: 'pkcs8',
        format: 'pem'
    });

    const publicKeyPem = keyPair.publicKey.export({
        type: 'spki',
        format: 'pem'
    });

    await ensureParentDirectoryExists(keyFilePath);

    // Save private key
    await fs.writeFile(keyFilePath, privateKeyPem);
    
    // Save public key alongside private key
    const publicKeyPath = `${keyFilePath}.pub`;
    await fs.writeFile(publicKeyPath, publicKeyPem);
}

/**
 * Load a private key from a file
 * 
 * @param keyFilePath Path to the private key file
 * @returns The loaded private key or null if the file doesn't exist
 */
export async function loadPrivateKey(keyFilePath: string): Promise<KeyObject | null> {
    try {
        if (await pathExists(keyFilePath)) {
            const privateKeyPem = await fs.readFile(keyFilePath, 'utf8');
            return createPrivateKey(privateKeyPem);
        }
        return null;
    } catch (error) {
        console.error(`Error loading private key from ${keyFilePath}:`, error);
        return null;
    }
}

/**
 * Load a public key from a file
 * 
 * @param keyFilePath Path to the public key file (typically ends with .pub)
 * @returns The loaded public key or null if the file doesn't exist
 */
export async function loadPublicKey(keyFilePath: string): Promise<KeyObject | null> {
    try {
        if (await pathExists(keyFilePath)) {
            const publicKeyPem = await fs.readFile(keyFilePath, 'utf8');
            return createPublicKey(publicKeyPem);
        }
        return null;
    } catch (error) {
        console.error(`Error loading public key from ${keyFilePath}:`, error);
        return null;
    }
}

/**
 * Loads or generates a key pair
 * 
 * @param keyFilePath Path to the private key file
 * @param generate Whether to generate a new key pair if none exists
 * @returns The key pair or null if no keys exist and generation is disabled
 */
export async function loadOrGenerateKeyPair(keyFilePath: string, generate = false): Promise<IKeyPair | null> {
    // First, try to load existing keys
    const privateKey = await loadPrivateKey(keyFilePath);
    const publicKey = await loadPublicKey(`${keyFilePath}.pub`);
    
    if (privateKey && publicKey) {
        return { privateKey, publicKey };
    }
    
    // If keys don't exist and generation is enabled, create new ones
    if (generate) {
        const keyPair = generateKeyPair();
        await saveKeyPair(keyPair, keyFilePath);
        return keyPair;
    }
    
    return null;
}

/**
 * Returns a 32-byte SHA-256 hash of the public key (SPKI format).
 * Used in the encrypted file header to identify which key encrypted the file.
 */
export function hashPublicKey(publicKey: KeyObject): Buffer {
    const spki = publicKey.export({ type: 'spki', format: 'der' }) as Buffer;
    return createHash('sha256').update(spki).digest();
}

/**
 * Load encryption keys for storage
 *
 * @param keyPaths Array of paths to key files (first is default/write key)
 * @param generateKey Whether to generate a key if it doesn't exist
 * @returns Storage options with encryption keys, or empty object if no key provided
 */
export async function loadEncryptionKeys(
    keyPaths: string[], 
    generateKey: boolean
): Promise<{ options: IStorageOptions, isEncrypted: boolean }> {
    if (!keyPaths || keyPaths.length === 0) {
        return { options: {}, isEncrypted: false };
    }
    
    const decryptionKeyMap: IPrivateKeyMap = {};
    let encryptionPublicKey: KeyObject | undefined = undefined;

    for (let index = 0; index < keyPaths.length; index++) {
        const keyPath = keyPaths[index];

        if (generateKey) {
            const keyPair = await loadOrGenerateKeyPair(keyPath, true);
            
            if (!keyPair) {
                throw new Error(`Failed to generate key pair at ${keyPath}`);
            }

            const keyHashHex = hashPublicKey(keyPair.publicKey).toString('hex');
            decryptionKeyMap[keyHashHex] = keyPair.privateKey;

            if (index === 0) {
                decryptionKeyMap.default = keyPair.privateKey;
                encryptionPublicKey = keyPair.publicKey;
            }
        } 
        else {
            const privateKey = await loadPrivateKey(keyPath);
            if (!privateKey) {
                throw new FatalError(`Private key not found: ${keyPath}\nUse --generate-key to create a new key or specify an existing key file either in the current directory or in the ~/.config/photosphere/keys directory.`);
            }
            
            let publicKey = await loadPublicKey(`${keyPath}.pub`);
            if (!publicKey) {
                publicKey = privateKey;
            }

            const keyHashHex = hashPublicKey(publicKey).toString('hex');
            decryptionKeyMap[keyHashHex] = privateKey;

            if (index === 0) {
                decryptionKeyMap.default = privateKey;
                encryptionPublicKey = publicKey;
            }
        }
    }

    if (!encryptionPublicKey || !decryptionKeyMap.default) {
        return { options: {}, isEncrypted: false };
    }

    return {
        options: {
            decryptionKeyMap,
            encryptionPublicKey
        },
        isEncrypted: true
    };
}

File: packages-zig/encryption-zig/src/lib/key-utils.zig
const std = @import("std");
const crypto = @import("node-crypto.zig");
const encryption_types = @import("encryption-types.zig");

//
// Names used from other files (the equivalent of the TypeScript imports).
//
const IStorageOptions = encryption_types.IStorageOptions;
const IPrivateKeyMap = encryption_types.IPrivateKeyMap;

//
// Interface for key pair
//
pub const IKeyPair = struct {
    // The public key.
    publicKey: *const crypto.PublicKey,

    // The private key.
    privateKey: *const crypto.PrivateKey,
};

//
// An RSA key pair stored as PEM strings.
//
pub const IEncryptionKeyPem = struct {
    // PEM-encoded PKCS#8 private key.
    privateKeyPem: []const u8,

    // PEM-encoded SPKI public key.
    publicKeyPem: []const u8,
};

//
// The result of loadEncryptionKeysFromPem (TypeScript: { options: IStorageOptions, isEncrypted: boolean }).
//
pub const ILoadedEncryptionKeys = struct {
    // Storage options with encryption keys (empty when not encrypted).
    options: IStorageOptions,

    // True when at least one key was loaded.
    isEncrypted: bool,
};

//
// Generate a new RSA key pair
//
// @returns The generated key pair
//
pub fn generateKeyPair(allocator: std.mem.Allocator, io: std.Io) !IKeyPair {
    const generated = try crypto.generateKeyPairSync(allocator, io, 4096);

    return IKeyPair{
        .publicKey = try crypto.createPublicKey(allocator, generated.publicKey),
        .privateKey = try crypto.createPrivateKey(allocator, generated.privateKey),
    };
}

// Not ported: saveKeyPair (not reached by replicate or verify)
// Not ported: loadPrivateKey (not reached by replicate or verify)
// Not ported: loadPublicKey (not reached by replicate or verify)
// Not ported: loadOrGenerateKeyPair (not reached by replicate or verify)

//
// Export a public key to SPKI PEM string (same format as .pub files and .db/encryption.pub).
//
pub fn exportPublicKeyToPem(allocator: std.mem.Allocator, publicKey: *const crypto.PublicKey) ![]u8 {
    return crypto.exportPublicKey(allocator, publicKey, .pem);
}

//
// Returns a 32-byte SHA-256 hash of the public key (SPKI format).
// Used in the encrypted file header to identify which key encrypted the file.
//
pub fn hashPublicKey(allocator: std.mem.Allocator, publicKey: *const crypto.PublicKey) ![32]u8 {
    const spki = try crypto.exportPublicKey(allocator, publicKey, .der);
    var digest: [32]u8 = undefined;
    std.crypto.hash.sha2.Sha256.hash(spki, &digest, .{});
    return digest;
}

//
// Load encryption keys from PEM strings without touching the filesystem.
// Accepts an array of { privateKeyPem, publicKeyPem } objects.
// The first entry becomes the default encryption (write) key.
//
// @param keyPems Array of PEM key pair objects
// @returns Storage options with encryption keys, or empty object if array is empty
//
pub fn loadEncryptionKeysFromPem(allocator: std.mem.Allocator, keyPems: []const IEncryptionKeyPem) !ILoadedEncryptionKeys {
    if (keyPems.len == 0) {
        return ILoadedEncryptionKeys{ .options = .{}, .isEncrypted = false };
    }

    var decryptionKeyMap: IPrivateKeyMap = .empty;
    var encryptionPublicKey: ?*const crypto.PublicKey = null;

    for (keyPems, 0..) |keyPem, index| {
        const privateKey = try crypto.createPrivateKey(allocator, keyPem.privateKeyPem);
        const publicKey = try crypto.createPublicKey(allocator, keyPem.publicKeyPem);

        const keyHash = try hashPublicKey(allocator, publicKey);
        const keyHashHex = try allocator.dupe(u8, &std.fmt.bytesToHex(keyHash, .lower));
        try decryptionKeyMap.put(allocator, keyHashHex, privateKey);

        if (index == 0) {
            try decryptionKeyMap.put(allocator, "default", privateKey);
            encryptionPublicKey = publicKey;
        }
    }

    if (encryptionPublicKey == null or decryptionKeyMap.get("default") == null) {
        return ILoadedEncryptionKeys{ .options = .{}, .isEncrypted = false };
    }

    return ILoadedEncryptionKeys{
        .options = .{
            .decryptionKeyMap = decryptionKeyMap,
            .encryptionPublicKey = encryptionPublicKey,
        },
        .isEncrypted = true,
    };
}

// Not ported: loadEncryptionKeys (not reached by replicate or verify; the CLI loads keys from the vault as PEM)
